src/atom_space_node/AtomSpaceNode.h

```diff
@@ -0,0 +1,193 @@
+
+using namespace std;
+
+namespace atom_space_node {
+
+/**
+ * Implements a node in a network used for some distributed processing.
+ *
+ * This is an abstract class so users of this module are supposed to extend it to
+ * implement its own distributed network.
+ *
+ * Communication between nodes are performed by Message objects which are like
+ * a piece of code sent from one node to another. We don't actually send
+ * code but instead we send a reference to what type of Message object is being
+ * exchanged so once the message arrive in the recipient a Message object is built
+ * and its code is executed. The "reference" about what type of Message is being
+ * exchanged is encoded like command lines with commands and its arguments.
+ *
+ * This is what you should care about when extending AtomSpaceNode:
+ *
+ *   - AtomSpaceNode is the class that build Message objects. That's why it extends
+ *     MessageFactory. This abstract class already knows how to build some basic Message
+ *     objects for some commands which are common to all the networks (e.g. joining
+ *     of new nodes to the network) but subclasses of AtomSpaceNode should re-implement
+ *     message_factory() to build Message objects for its own distributed application.
+ *   - Message commands are executed in separated threads. So if such commands updates some
+ *     state variable inside AtomSpaceNode or its subclasses, this state variable should
+ *     be protected by mutual exclusion primitives.
+ *   - AtomSpaceNode's constructor expects allocated objects for MessageBroker and
+ *     LeadershipBroker. These are abstract classes which may have many concrete
+ *     implementations. When designing your distributed network you need to decide
+ *     which messaging system is supposed to be used to actually make the communication
+ *     bwtween nodes. You can use one of the concrete implementations we provide or you
+ *     can provide your own implementation extending these abstract classes. Typically, users
+ *     wouldn't need to implement its own messaging system but it may be the case for the
+ *     leadership election algorithm as this is highly dependent on the network topology
+ *     and expected behaviour.
+ *   - AtomSpaceNode has some pure virtual methods that need to be implemented. These are
+ *     the methods called by basic Message objects to acomplish basic tasks like leadership
+ *     election and notification of new nodes joning the network. The proper way to respond
+ *     to these requests are delegated to concrete classes extending AtomSpaceNode.
+ */
+class AtomSpaceNode : public MessageFactory {
+
+public:
+
+    /**
+     * Destructor.
+     *
+     * TODO: Implement a graceful shutdown of all components
+     */
+    ~AtomSpaceNode();
+
+    // --------------------------------------------------------------------------------------------
+    // Public API
+
+    /**
+     * Joins a network of similar nodes.
+     */
+    void join_network();
```

Comment:
As the node considers itself joined once the network has an elected leader, I think this doxygen should emphasize that and highlight that there is no validation or approval from a master node (or similar concept) or from the other nodes in the network.

Reply:
Maybe this details should be in the implementation file (`.cc`) instead.

Reply:
Added this to the scope of https://github.com/singnet/das/issues/99